contribs/drt/src/main/java/org/matsim/contrib/drt/speedup/DrtSpeedUp.java

```diff
@@ -0,0 +1,240 @@
+		SimulatedTripStats tripStats = computeSimulatedTripStats();
+		log.info("Number of simulated " + mode + " trips: " + tripStats.count);
+
+		// store additional information
+		averageWaitingTimes.add(tripStats.averageWaitTime);
+		averageInVehicleBeelineSpeeds.add(tripStats.averageInVehicleBeelineSpeed);
+
+		double movingAverageWaitingTime = computeMovingAverage(drtSpeedUpParams.getMovingAverageSize(),
+				averageWaitingTimes);
+		log.info("Setting waiting time for {} to: {} (previous value: {})", mode, movingAverageWaitingTime,
+				currentAvgWaitingTime);
+		currentAvgWaitingTime = movingAverageWaitingTime;
+
+		double movingAverageInVehicleBeelineSpeed = computeMovingAverage(drtSpeedUpParams.getMovingAverageSize(),
+				averageInVehicleBeelineSpeeds);
+		log.info("Setting in-vehicle beeline speed for {} to: {} (previous value: {})", mode,
+				movingAverageInVehicleBeelineSpeed, currentAvgInVehicleBeelineSpeed);
+		currentAvgInVehicleBeelineSpeed = movingAverageInVehicleBeelineSpeed;
+
+		if (drtSpeedUpParams.getWaitingTimeUpdateDuringSpeedUp()
+				== DrtSpeedUpParams.WaitingTimeUpdateDuringSpeedUp.LinearRegression) {
+			// store some additional statistics
+			double fleetSize = fleetSpecification.getVehicleSpecifications().size();
+			Preconditions.checkState(fleetSize < 1, "No vehicles for drt mode %s. Aborting...", mode);
+			double ridesPerVehicle = tripStats.count / fleetSize;
+			ridesPerVehicle2avgWaitingTime.add(new Tuple<>(ridesPerVehicle, currentAvgWaitingTime));
+		}
+	}
+
+	private static class SimulatedTripStats {
+		private final int count;
+		private final double averageInVehicleBeelineSpeed;
+		private final double averageWaitTime;
+
+		private SimulatedTripStats(int count, double averageInVehicleBeelineSpeed, double averageWaitTime) {
+			this.count = count;
+			this.averageInVehicleBeelineSpeed = averageInVehicleBeelineSpeed;
+			this.averageWaitTime = averageWaitTime;
+		}
+	}
+
+	private SimulatedTripStats computeSimulatedTripStats() {
+		Mean meanInVehicleBeelineSpeed = new Mean();
+		Mean meanWaitTime = new Mean();
+
+		for (var sequence : drtRequestAnalyzer.getPerformedRequestSequences().values()) {
+			if (!sequence.isCompleted()) {
+				continue;//skip incomplete sequences
+			}
+			DrtRequestSubmittedEvent submittedEvent = sequence.getSubmitted();
+
+			Link depLink = network.getLinks().get(submittedEvent.getFromLinkId());
+			Link arrLink = network.getLinks().get(submittedEvent.getToLinkId());
+			double beelineDistance = DistanceUtils.calculateDistance(depLink, arrLink);
+
+			double pickupTime = sequence.getPickedUp().get().getTime();
+			double waitTime = pickupTime - sequence.getSubmitted().getTime();
+			double rideTime = sequence.getDroppedOff().get().getTime() - pickupTime;
+
+			//TODO do we need these checks?
+			Preconditions.checkState(rideTime > 0);
+			Preconditions.checkState(beelineDistance > 0);
+			//TODO I would map unshared_ride_time to rideTime -- should be more precise
```

Comment:
`DrtRouteUpdater` updates all routes before each iteration. It's highly parallelisable, so does not cost much, but helps being more accurate about max travel & wait times constraints in the DRT insertion algorithm. Of course these are **expected** "unshared" travel times and distances.

I think it is okay if drt-speed-up maintains a mapping: **expected_unshared_TT -> actual_shared_TT**. This is actually what the simulated DRT also does but in a less direct way.

Moreover, DRT (DVRP, in general) uses exponentially-averaged-over-iterations TTs for actual vehicle routing (to avoid oscillations...). The same TTs are used by `DrtRouteUpdater`. So I hope the trickiest part is already solved :-)

I will leave this TODO for the future. I think it is worth exploring. My intuition says that **expected_unshared_TT -> actual_shared_TT** should be more precise (or at least closer to the decision-making process on the optimiser side).

Reply:
That makes sense. I also think that **expected_unshared_TT -> actual_shared_TT** should be more accurate than the current approach.